### PR TITLE
1400 define particle filter

### PR DESCRIPTION
**Pull request by Bogdan2423**
obraz

### CODE REVIEW

**Comment by Bogdan2423** [comment]
* In current version of project there is a bug which makes it impossible to remove filters. I didnt manage to fix it yet. I would prefer to create a seperate issue for the bug and fix it there, unless you think that I should fix it in this PR
* Adding, duplicating, loading and saving to JSON and adding filters to detectors work normally for both types of filters
* This change is not backwards compatible, as SHIELDHIT-12A converter cannot parse the new JSON with particle filter.
The new JSON format:
`"filters": [
			{
				"uuid": "c0e0f38d-4f19-4d80-8f63-0e955059a96b",
				"name": "Filter",
				"type": "Filter",
				"rules": [
					{
						"uuid": "af471dee-8dde-4eaa-8a1d-1c26097f7336",
						"keyword": "Z",
						"operator": "equal",
						"value": 1
					}]
			},
			{
				"uuid": "1c47cbc8-29f2-4eb0-b275-4876c331d9e2",
				"name": "Filter_1",
				"type": "Filter",
				"particle": {
					"id": 2,
					"name": "Proton"
				}
			},
...`
For the new particle filter the rules field is replaced with "particle" field which has fields "id" and "name". The rest stays the same.

**Comment by grzanka** [comment]
> * In current production version of project there is a bug which makes it impossible to remove filters. I didnt manage to fix it yet. I would prefer to create a seperate issue for the bug and fix it there, unless you think that I should fix it in this PR

Thats not a problem, it can go into separate PR / issue.

> * Adding, duplicating, loading and saving to JSON and adding filters to detectors work normally for both types of filters

Good

> * This change is not backwards compatible, as SHIELDHIT-12A converter cannot parse the new JSON with particle filter.
>   The new JSON format:
>   `"filters": [ { "uuid": "c0e0f38d-4f19-4d80-8f63-0e955059a96b", "name": "Filter", "type": "Filter", "rules": [ { "uuid": "af471dee-8dde-4eaa-8a1d-1c26097f7336", "keyword": "Z", "operator": "equal", "value": 1 }] }, { "uuid": "1c47cbc8-29f2-4eb0-b275-4876c331d9e2", "name": "Filter_1", "type": "Filter", "particle": { "id": 2, "name": "Proton" } }, ...`
>   For the new particle filter the rules field is replaced with "particle" field which has fields "id" and "name". The rest stays the same.

Then maybe you should bump JSON version ? What will happen if you define "Custom filter" (one based on Rules) and run now SHIELD-HIT12A simulation ? Will converter accept that ?

File: src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { Object3D } from 'three';
```
Comment (grzanka) [inline src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx:1]: Do we need to have a link to 3D editor here ?

File: src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import { SetValueCommand } from '../../../../js/commands/SetValueCommand';
+import { YaptideEditor } from '../../../../js/YaptideEditor';
+import { isParticleFilter, ParticleFilter } from '../../../../Simulation/Scoring/ParticleFilter';
+import { ParticleSelect } from '../../../Select/ParticleSelect';
+import { PropertiesCategory } from './PropertiesCategory';
+
+export function ParticleFilterConfiguration(props: { editor: YaptideEditor; object: Object3D }) {
+	const { object, editor } = props;
+
+	const { state: watchedObject } = useSmartWatchEditorState(editor, object as ParticleFilter);
```
Comment (grzanka) [inline src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx:14]: do we need to use smartwatcheditor and monitor 3d editor here ?

**Review by grzanka — APPROVED** [review]

File: src/ThreeEditor/Simulation/Scoring/ScoringManager.ts [hunk 3]
```diff
@@ -110,7 +117,9 @@ export class ScoringManager
 	getFilterOptions(): Record<string, string> {
 		const options = this.filters
 			.filter(filter => {
-				return filter.rules.length;
+				if (filter instanceof CustomFilter) return filter.rules.length;
+
+				return true;
```
Comment (hendzeld) [inline src/ThreeEditor/Simulation/Scoring/ScoringManager.ts:122]: Simplified using ternary operator.

```suggestion
			.filter(filter => filter instanceof CustomFilter ? filter.rules.length : true)
```

File: src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx (hunk 2 above)
Comment (Bogdan2423) [reply src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx:14]: This is how all other "configuration" elements keep track of state of related object. Currently I dont know if there is another way to do it, and I dont think there is a reason to do it differently.

File: src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx (hunk 1 above)
Comment (Bogdan2423) [reply src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx:1]: We need if we are going to use smartwatcheditor

**Comment by Bogdan2423** [comment]
> > * This change is not backwards compatible, as SHIELDHIT-12A converter cannot parse the new JSON with particle filter.
> >   The new JSON format:
> >   `"filters": [ { "uuid": "c0e0f38d-4f19-4d80-8f63-0e955059a96b", "name": "Filter", "type": "Filter", "rules": [ { "uuid": "af471dee-8dde-4eaa-8a1d-1c26097f7336", "keyword": "Z", "operator": "equal", "value": 1 }] }, { "uuid": "1c47cbc8-29f2-4eb0-b275-4876c331d9e2", "name": "Filter_1", "type": "Filter", "particle": { "id": 2, "name": "Proton" } }, ...`
> >   For the new particle filter the rules field is replaced with "particle" field which has fields "id" and "name". The rest stays the same.
> 
> Then maybe you should bump JSON version ? What will happen if you define "Custom filter" (one based on Rules) and run now SHIELD-HIT12A simulation ? Will converter accept that ?

I bumped the version. If there is no new "particle filter" defined then converter will work normally and you can use "custom filter". It only will not work if you try to run simulation with the new type of filter.

File: src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx (hunk 2 above)
Comment (grzanka) [reply src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx:14]: OK, then lets keep it

File: src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx (hunk 1 above)
Comment (grzanka) [reply src/ThreeEditor/components/Sidebar/properties/category/ParticleFilterConfiguration.tsx:1]: OK

**Review by grzanka — APPROVED** [review]

**Pull request merged**